lib/pacemaker/pcmk_sched_group.c

```diff
@@ -29,12 +29,46 @@ expand_group_colocations(pe_resource_t *rsc)
 {
     group_variant_data_t *group_data = NULL;
     pe_resource_t *member = NULL;
+    bool any_unmanaged = false;
 
     get_group_variant_data(group_data, rsc);
 
     // Treat "group with R" colocations as "first member with R"
     member = group_data->first_child;
     member->rsc_cons = g_list_concat(member->rsc_cons, rsc->rsc_cons);
+
+
+    /* The above works for the whole group because each group member is
+     * colocated with the previous one.
+     *
+     * However, there is a special case when a group has a mandatory colocation
+     * with a resource that can't start. In that case, update_colo_start_chain()
+     * will ensure that dependent resources in mandatory colocations (i.e. the
+     * first member for groups) can't start either. But if any group member is
+     * unmanaged and already started, the internal group colocations are no
+     * longer sufficient to make that apply to later members.
+     *
+     * To handle that case, add mandatory colocations to each member after the
+     * first.
+     */
+    any_unmanaged = !pcmk_is_set(member->flags, pe_rsc_managed);
+    for (GList *item = rsc->children->next; item != NULL; item = item->next) {
+        member = item->data;
```

Comment:
Probably I'm not getting this properly.
My feeling would have been to rather walk over the members of the group.

Reply:
That's what it does, excluding the first, since the first is already handled by the previous lines.

Basically the question is how to implement "colocate group with R".

One way would be to colocate each group member individually with R.

What we did instead was colocate just the first member with R. Since each group member is colocated with the one before it, that is generally sufficient.

However, if any group members are unmanaged, it will not be sufficient if an unmanaged member is already started on some other node. In that case, members after the unmanaged member will be colocated with that member, not R. If the colocation with R is optional, then we can just ignore it, but if it's mandatory, we should explicitly colocate each later member with R, which is what this does. If the unmanaged member happens to be on the node with R, everything will be fine, and otherwise, the inability to meet both the mandatory colocation and the internal group colocation will make later members stop.